critical_section.c

```diff
@@ -0,0 +1,24 @@
+#include <critical_section.h>
+#include <interrupts.h>
+#include <common.h>
+#include <libkern.h>
+
+volatile int cs_level = 0;
```

Comment:
That's why interrupt disable and enable should also be a compiler barrier (as I mentioned in another comment) so that compiler won't misoptimise things. Right, now `cs_level` is indeed protected but other memory accesses in the critical section can still be reordered.

Counter for interrupt disable/enable seems like a good idea, it's similar to that saving previous state on interrupt disable I mentioned the other time.